library/test/src/helpers/shuffle.rs

```diff
@@ -0,0 +1,67 @@
+use crate::cli::TestOpts;
+use crate::types::{TestDescAndFn, TestId, TestName};
+use std::collections::hash_map::DefaultHasher;
+use std::hash::Hasher;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+pub fn get_shuffle_seed(opts: &TestOpts) -> Option<u64> {
+    opts.shuffle_seed.or_else(|| {
+        if opts.shuffle {
+            Some(
+                SystemTime::now()
+                    .duration_since(UNIX_EPOCH.into())
+                    .expect("Failed to get system time")
+                    .as_nanos() as u64,
+            )
+        } else {
+            None
+        }
+    })
+}
+
+pub fn shuffle_tests(shuffle_seed: u64, tests: &mut [(TestId, TestDescAndFn)]) {
+    let test_names: Vec<&TestName> = tests.iter().map(|test| &test.1.desc.name).collect();
+    let test_names_hash = calculate_hash(&test_names);
+    let mut rng = Rng::new(shuffle_seed, test_names_hash);
+    shuffle(&mut rng, tests);
+}
+
+// `shuffle` is from `rust-analyzer/src/cli/analysis_stats.rs`.
+fn shuffle<T>(rng: &mut Rng, slice: &mut [T]) {
+    for i in 0..slice.len() {
+        randomize_first(rng, &mut slice[i..]);
+    }
+
+    fn randomize_first<T>(rng: &mut Rng, slice: &mut [T]) {
+        assert!(!slice.is_empty());
+        let idx = rng.rand_range(0..slice.len() as u64) as usize;
+        slice.swap(0, idx);
+    }
+}
+
+struct Rng {
+    state: u64,
+    extra: u64,
+}
+
+impl Rng {
+    fn new(seed: u64, extra: u64) -> Self {
+        Self { state: seed, extra }
+    }
+
+    fn rand_range(&mut self, range: core::ops::Range<u64>) -> u64 {
+        self.rand_u64() % (range.end - range.start) + range.start
+    }
+
+    fn rand_u64(&mut self) -> u64 {
+        self.state = calculate_hash(&(self.state, self.extra));
```

Comment:
can we use a proper RNG library like `fastrand` or `oorandom` or `rand`

there is no proof that `state = sip13_hash((state, extra))` is a good RNG.

(and additionally `rand_u64() % n` produces  biased result)

Reply:
> can we use a proper RNG library like `fastrand` or `oorandom` or `rand`

I tried `rand` and `oorandom` actually. In both cases, I got `can't find crate for core` errors.

My impression (based on [this](https://github.com/rust-lang/rust/blob/0a0688e6f51229d2860af6317adaef378f4b45a5/library/test/Cargo.toml#L10-L11), [this](https://github.com/alexcrichton/cfg-if/blob/59f5b77e6255ce8baf304f74f141a67d42f339c5/Cargo.toml#L21), and [this](https://github.com/rust-lang/getopts/blob/c11eb65097d0dc9fe9ceb351d63c3ac8b5f05e0b/Cargo.toml#L19)) is that a crate has to go out of its way to be a dependency of a builtin library. And none of those crates seems to do that.

Note that `oorandom` barely uses `core`, but it does use it for `Range` ([here](https://hg.sr.ht/~icefox/oorandom/browse/src/lib.rs?rev=tip#L23)).

> there is no proof that `state = sip13_hash((state, extra))` is a good RNG.
> 
> (and additionally `rand_u64() % n` produces biased result)

Your points are very well taken. However, I don't think this random number generator needs to be "strong." Put another way, I think the risk of an attacker trying to manipulate the test ordering is low. I am open to counterarguments though.

(Thanks for doing this, BTW.)